packages/flutter_tools/lib/src/android/android_device.dart

```diff
@@ -263,18 +263,28 @@ class AndroidDevice extends Device {
   }
 
   bool _installLatestApp(ApplicationPackage package) {
-    if (isAppInstalled(package)) {
+    bool wasInstalled = isAppInstalled(package);
+    if (wasInstalled) {
       if (isLatestBuildInstalled(package)) {
         printStatus('Latest build already installed.');
         return true;
       }
-      printStatus('Uninstalling old version...');
-      if (!uninstallApp(package))
-        printError('Warning: uninstalling old version failed');
     }
     printTrace('Installing APK.');
     if (!installApp(package)) {
       printTrace('Error: Failed to install APK.');
+      if (wasInstalled) {
+        printStatus('Uninstalling old version...');
+        if (!uninstallApp(package)) {
+          printError('Error: Uninstalling old version failed.');
+          return false;
+        }
+        if (!installApp(package)) {
+          printTrace('Error: Failed to install APK again.');
```

Comment:
You want printError for anything fatal, printStatus for anything critical, and printTrace for debugging information.

I think in the case you're talking about, it's debugging (because if the next step works, then the user need not know about any of it, and it's not fatal), so I'd stick to printTrace.